public/app/features/dashboard/state/PanelModel.ts

```diff
@@ -269,6 +270,33 @@ export class PanelModel implements DataConfigSource, IPanelModel {
     }
   }
 
+  // Attempts to populate each panel target with a well-formed query definition
+  async ensureWellFormedQueries() {
+    if (this.targets && isArray(this.targets)) {
+      // If we modified any targets, publish one update event
+      let shouldPublish = false;
+      for (let i = 0; i < this.targets.length; i++) {
+        const query = this.targets[i];
+        if (!query.datasource && this.datasource) {
+          query.datasource = this.datasource;
+        }
+        if (Object.keys(query).length === 2) {
+          // If query has only refId and datasource, try to set the default query for its datasource
+          const datasourceInstance = await getDataSourceSrv().get(query.datasource);
+          const defaultQuery = datasourceInstance.getDefaultQuery?.(CoreApp.PanelEditor);
+          if (defaultQuery) {
+            this.targets[i] = { ...defaultQuery, ...query };
+            shouldPublish = true;
+          }
+        }
+      }
+      if (shouldPublish) {
+        this.configRev++;
+        this.events.publish(new PanelQueriesChangedEvent());
+      }
```

Comment:
not sure issuing queries should cause modifications to the model

Reply:
roger